kv/etcd/mock.go

```diff
@@ -1,71 +1,480 @@
+// is determined to be an acceptable tradeoff to avoid needing to depend on an entire
+// Etcd server for unit tests.
+//
+// Known limitations:
+//
+// * Compact is not implemented and will panic
+// * RequestProgress is not implemented and will panic
+// * Only exact and prefix matching is supported for Get, Put, and Delete
+// * There may be inconsistencies with how various version numbers are adjusted
+//   but none that are exposed by kv.Client unit tests
+type mockKV struct {
+	// Key-value pairs created by put calls or transactions
+	values    map[string]mvccpb.KeyValue
+	valuesMtx sync.Mutex
+
+	// Channel for stopping all running watch goroutines and closing
+	// and cleaning up all channels used for sending events to watchers
+	close chan struct{}
+
+	// Channels that should receive events in response to Put or Delete
+	// calls. These channels are in turn read by goroutines that apply
+	// filtering before sending watch responses to their callers.
+	consumers    map[chan clientv3.Event]struct{}
+	consumersMtx sync.Mutex
+}
+
+// Watch implements the Clientv3Facade interface
+func (m *mockKV) Watch(ctx context.Context, key string, opts ...clientv3.OpOption) clientv3.WatchChan {
+	out := make(chan clientv3.WatchResponse)
+	consumer := m.createConsumer()
+
+	go func() {
+		defer func() {
+			// When this goroutine ends, remove and close the channel written to by the
+			// Put and Delete methods as well as closing the channel read by the caller
+			// of the Watch method
+			m.destroyConsumer(consumer)
+			out <- clientv3.WatchResponse{Canceled: true}
+			close(out)
+		}()
+
+		for {
+			select {
+			case <-ctx.Done():
+				// Context cancelled for this watcher, run cleanup logic and exit
+				return
+			case <-m.close:
+				// Close method called for all watchers, run cleanup logic and exit
+				return
+			case e := <-consumer:
+				op := clientv3.OpGet(key, opts...)
+				match := m.isMatch(op, *e.Kv)
+
+				if match {
+					out <- clientv3.WatchResponse{Events: []*clientv3.Event{&e}}
+				}
+			}
+		}
+	}()
+
+	return out
+}
+
+// createConsumer creates and returns a new channel that is registered to receive
+// events for Puts and Deletes.
+func (m *mockKV) createConsumer() chan clientv3.Event {
+	ch := make(chan clientv3.Event)
+	m.consumersMtx.Lock()
+	m.consumers[ch] = struct{}{}
+	m.consumersMtx.Unlock()
+	return ch
+}
+
+// destroyConsumer removes the given channel from the list of channels that events
+// should be sent to and closes it.
+func (m *mockKV) destroyConsumer(ch chan clientv3.Event) {
+	m.consumersMtx.Lock()
+	delete(m.consumers, ch)
+	m.consumersMtx.Unlock()
+	close(ch)
+}
+
+// sendEvent writes an event to all currently registered consumers. The consumer
+// channels are each read by a goroutine that filters the event and sends it to
+// the caller of the Watch method.
+func (m *mockKV) sendEvent(e clientv3.Event) {
+	m.consumersMtx.Lock()
+	for ch := range m.consumers {
+		ch <- e
```

Comment:
should this be a non-blocking write? the channels created in `createConsumer` have a buffer of `0`, so a bad consumer could block this.

Reply:
Good idea. I think it's probably a good idea to use channels with a non-zero buffer as well. WDYT?

Reply:
Fixed in c8424f3ad074264a691b08ec6e1bb3aa1a9e54f2

Reply:
> I think it's probably a good idea to use channels with a non-zero buffer as well. WDYT?

yes, definitely